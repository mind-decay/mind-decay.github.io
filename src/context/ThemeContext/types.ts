import { Nullable } from 'types';

export type ThemeType = 'light' | 'dark';

export type ThemeContextType = {
  theme: ThemeType;
  toggleTheme: () => void;
};

export const isThemeType = (theme: Nullable<string>): theme is ThemeType => {
  return typeof theme === 'string' && ['light', 'dark'].includes(theme);
};
