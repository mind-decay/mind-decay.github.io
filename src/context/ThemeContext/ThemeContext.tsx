import { ReactNode, createContext, useContext, useEffect, useState } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';

import { ThemeContextType, ThemeType, isThemeType } from './types';
import { Nullable } from 'types/helpers';

import { darkTheme, lightTheme } from 'styles/themes';

const ThemeContext = createContext<Nullable<ThemeContextType>>(null);

export const useTheme = (): ThemeContextType => {
  const context = useContext(ThemeContext);

  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }

  return context;
};

type ThemeProviderProps = {
  children: ReactNode;
};

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [theme, setTheme] = useState<ThemeType>('dark');
  const isLightTheme = theme === 'light';

  const toggleTheme = () => {
    const updatedTheme = isLightTheme ? 'dark' : 'light';
    setTheme(updatedTheme);
    localStorage.setItem('theme', updatedTheme);
  };

  useEffect(() => {
    const savedTheme = localStorage.getItem('theme');
    const prefersLight = window.matchMedia && window.matchMedia('(prefers-color-scheme: light)').matches;
    const checkCurrentTheme = isThemeType(savedTheme);

    if (checkCurrentTheme) {
      setTheme(savedTheme);
    } else if (prefersLight) {
      setTheme('light');
    }
  }, []);

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      <StyledThemeProvider theme={isLightTheme ? lightTheme : darkTheme}>{children}</StyledThemeProvider>
    </ThemeContext.Provider>
  );
};
