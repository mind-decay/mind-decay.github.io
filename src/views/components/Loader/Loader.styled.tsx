import styled, { keyframes } from 'styled-components';

const animation = keyframes`
  to {
      background-size: 100% 3px;
    }
`;

export const LoaderWrapper = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1000;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  background-color: ${({ theme }) => theme.colors.secondary};
`;

export const StyledLoader = styled.div`
  width: fit-content;
  font-weight: bold;
  font-family: sans-serif;
  font-size: 30px;
  padding-bottom: 8px;
  background: ${({ theme }) => `linear-gradient(${theme.colors.primary} 0 0) 0 100%/0% 3px no-repeat`};
  animation: ${animation} 2s linear infinite;

  &::before {
    content: 'Loading...';
    color: ${({ theme }) => theme.colors.primary};
  }
`;
