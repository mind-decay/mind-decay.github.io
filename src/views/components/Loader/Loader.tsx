import { LoaderWrapper, StyledLoader } from './Loader.styled';

export const Loader = () => {
  return (
    <LoaderWrapper>
      <StyledLoader data-testid="loader" />
    </LoaderWrapper>
  );
};
