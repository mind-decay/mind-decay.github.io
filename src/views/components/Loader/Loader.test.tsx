import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import { ThemeProvider } from 'context/ThemeContext';

import { Loader } from './Loader';

test('render Loader component', () => {
  const { getByTestId } = render(
    <ThemeProvider>
      <Loader />
    </ThemeProvider>,
  );

  const loader = getByTestId('loader');

  expect(loader).toBeInTheDocument();
  expect(loader).toMatchSnapshot();

  expect(loader).toHaveAttribute('data-testid', 'loader');
});
