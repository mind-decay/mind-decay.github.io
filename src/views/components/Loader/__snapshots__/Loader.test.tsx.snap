// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render Loader component 1`] = `
<div
  class="sc-guDLey lagWuR"
  data-testid="loader"
/>
`;
